sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpEventHubProducer.cs

```diff
@@ -0,0 +1,417 @@
+                        // Otherwise, bubble the exception.
+
+                        ++failedAttemptCount;
+                        retryDelay = _retryPolicy.CalculateRetryDelay(ex, failedAttemptCount);
+
+                        if ((retryDelay.HasValue) && (!ConnectionScope.IsDisposed))
+                        {
+                            cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
+
+                            await Task.Delay(retryDelay.Value, cancellationToken).ConfigureAwait(false);
+                            stopWatch.Reset();
+                        }
+                        else
+                        {
+                            throw;
+                        }
+                    }
+                }
+
+                // If no value has been returned nor exception thrown by this point,
+                // then cancellation has been requested.
+
+                throw new TaskCanceledException();
+            }
+            finally
+            {
+                stopWatch.Stop();
+                EventHubsEventSource.Log.EventPublishComplete(EventHubName, logPartition, messageHash);
+            }
+        }
+
+        /// <summary>
+        ///   Creates the AMQP link to be used for producer-related operations and ensures
+        ///   that the corresponding state for the producer has been updated based on the link
+        ///   configuration.
+        /// </summary>
+        ///
+        /// <param name="partitionId">The identifier of the Event Hub partition to which it is bound; if unbound, <c>null</c>.</param>
+        /// <param name="timeout">The timeout to apply when creating the link.</param>
+        /// <param name="cancellationToken">The cancellation token to consider when creating the link.</param>
+        ///
+        /// <returns>The AMQP link to use for producer-related operations.</returns>
+        ///
+        /// <remarks>
+        ///   This method will modify class-level state, setting those attributes that depend on the AMQP
+        ///   link configuration.  There exists a benign race condition in doing so, as there may be multiple
+        ///   concurrent callers.  In this case, the attributes may be set multiple times but the resulting
+        ///   value will be the same.
+        /// </remarks>
+        ///
+        protected virtual async Task<SendingAmqpLink> CreateLinkAndEnsureProducerStateAsync(string partitionId,
+                                                                                            TimeSpan timeout,
+                                                                                            CancellationToken cancellationToken)
+        {
+            SendingAmqpLink link = await ConnectionScope.OpenProducerLinkAsync(partitionId, timeout, CancellationToken.None).ConfigureAwait(false);
+
+            if (!MaximumMessageSize.HasValue)
+            {
+                await Task.Delay(15).ConfigureAwait(false);
```

Comment:
We're not using the `CancellationToken` at all here. Shouldn't we pass it to `OpenProducerLinkAsync` and `Delay`?

Reply:
gah.   Yes.  That's a mistake.